src/jit/codegenxarch.cpp

```diff
@@ -7356,6 +7356,142 @@ void CodeGen::genSSE2BitwiseOp(GenTreePtr treeNode)
     inst_RV_RV(ins, targetReg, operandReg, targetType);
 }
 
+//-----------------------------------------------------------------------------------------
+// genSSE41RoundOp - generate SSE41 code for the given tree as a round operation
+//
+// Arguments:
+//    treeNode  - tree node
+//
+// Return value:
+//    None
+//
+// Assumptions:
+//     i) SSE4.1 is supported by the underlying hardware
+//    ii) tree oper is a GT_INTRINSIC
+//   iii) tree type is floating point type
+//    iv) tree oper is CORINFO_INTRINSIC_Round, _Ceiling, or _Floor
+//     v) caller of this routine needs to call genProduceReg()
+void CodeGen::genSSE41RoundOp(GenTreeOp* treeNode)
+{
+    // i) SSE4.1 is supported by the underlying hardware
+    assert(compiler->compSupports(InstructionSet_SSE41));
+
+    // ii) tree oper is a GT_INTRINSIC
+    assert(treeNode->OperGet() == GT_INTRINSIC);
+
+    GenTree* srcNode = treeNode->gtGetOp1();
+
+    // iii) tree type is floating point type
+    assert(varTypeIsFloating(srcNode));
+    assert(srcNode->TypeGet() == treeNode->TypeGet());
+
+    genConsumeOperands(treeNode);
+
+    instruction ins  = (treeNode->TypeGet() == TYP_FLOAT) ? INS_roundss : INS_roundsd;
+    emitAttr    size = emitTypeSize(treeNode);
+
+    regNumber dstReg = treeNode->gtRegNum;
+    regNumber srcReg = srcNode->gtRegNum;
+
+    unsigned ival = 0;
+
+    // iv) tree oper is CORINFO_INTRINSIC_Round, _Ceiling, or _Floor
+    switch (treeNode->gtIntrinsic.gtIntrinsicId)
+    {
+        case CORINFO_INTRINSIC_Round:
+            ival = 4;
+            break;
+
+        case CORINFO_INTRINSIC_Ceiling:
+            ival = 10;
+            break;
+
+        case CORINFO_INTRINSIC_Floor:
+            ival = 9;
+            break;
+
+        default:
+            ins = INS_invalid;
+            assert(!"genSSE41: unsupported oper");
+            unreached();
+            break;
+    }
+
+    if (srcNode->isContained())
+    {
+        emitter* emit   = getEmitter();
+
+        unsigned varNum = BAD_VAR_NUM;
+        unsigned offset = (unsigned)-1;
+
+        if (srcNode->isIndir())
+        {
+            GenTreeIndir* mem     = srcNode->AsIndir();
+            GenTree*      memBase = mem->gtOp1;
+
+            switch (memBase->OperGet())
+            {
+                case GT_CLS_VAR_ADDR:
+                {
+                    emit->emitIns_R_C_I(ins, size, dstReg, memBase->gtClsVar.gtClsVarHnd, 0, ival);
+                }
+                return;
+
+                case GT_LCL_VAR_ADDR:
+                {
+                    varNum = memBase->AsLclVarCommon()->GetLclNum();
+                    offset = 0;
+                }
+                break;
+
+                default:
```

Comment:
`emitInsBinary` handles `address mode` instructions here. I'm not sure if I need to cover that for `roundsd`, however.

Reply:
You probably need to. Try a `Math.Round(a[i])` or something like that.

Reply:
👍, will test locally and update if needed.

Reply:
Updated so that address mode instructions work -- this was actually a bit more 'involved' then I initially though, as `emitOutputAM` didn't have `Is4ByteAVXInstruction` support like `emitOutputCV` and `emitOutputSV`.